cmd/tink-worker/internal/registry.go

```diff
@@ -64,8 +74,17 @@ func (r *RegistryConnDetails) pullImage(ctx context.Context, cli *client.Client,
 		return errors.Wrap(err, "DOCKER PULL")
 	}
 	defer out.Close()
-	if _, err := io.Copy(os.Stdout, out); err != nil {
-		return err
+	fd := json.NewDecoder(out)
+	var status *ImagePullStatus
+	for {
+		if err := fd.Decode(&status); err != nil {
+			if err == io.EOF {
+				break
+			}
+		}
+		if status.Error != "" {
+			return errors.Wrap(errors.New(status.Error), "DOCKER PULL")
+		}
```

Comment:
This changes the behavior and now hides the pulling logs, which I guess might be helpful while pulling large image.

Reply:
yes, that would be only possible if we print the `status`.

Reply:
Maybe we can use [TeeReader](https://golang.org/pkg/io/#TeeReader) to both print the messages and decode them to retain the old behavior?

Reply:
so TeeReader will return a Reader object and data will be pushed to stdout when we read it. Which will fail but go won't catch the error. I tested it, while downloading the image it failed in between, it didn't return the error.

Reply:
I think we can defer io.Copy on tee reader to keep the output, then read and decode it as done in this PR. What do you think?

Reply:
I don't think defer will make a difference. Both of them eventually will call the `read`. Secondly, reader object can not be read twice. 
Even if we mae it work, IMHO that would be an overhead for the `image pull` method. 
Do you have any specific use case where logs are required ? 

I think, one of the below three an be an approach:
1) keep `io.Copy` and document the error. 
2) keep the code change as in PR. For logs, we can add a logger from a separate issue. 
3) File a issue in repo of docker. Actually, you will find the issue on docker repo with the same context, and suggestions are there are to increase the host size. I didn't got lucky to find some other suggestions.

Reply:
>Do you have any specific use case where logs are required ?

It's the progress indicator, it's always nice to see in logs that something is happening. But if maintainers agrees this is not very important, I'm fine with it.

>Secondly, reader object can not be read twice.

This is why you could use TeeReader, to duplicate the stream.

I also found https://github.com/moby/moby/issues/36253 now, it seems `ImagePull` has even more bugs :|

Reply:
> I also found moby/moby#36253 now, it seems ImagePull has even more bugs :|

Woh! this is surprising, I didn't notice this. 

> This is why you could use TeeReader, to duplicate the stream.

I tried to replicate this scenario, as in below code snippet, but I didn't observed that object can be streamed/read again. Can you share me an example on what you are suggesting, 

```	reader, err := cli.ImagePull(ctx, "docker.io/cbkhare/test", types.ImagePullOptions{}) //RegistryAuth: authStr})
	if err != nil {
		fmt.Println(reader)
		return
	}
	r := io.TeeReader(reader, os.Stdout)

        // streaming the object
	_, err= io.ReadAll(r)
        if err != nil {
                return err
        }

        // re streaming the object. 
        fd := json.NewDecoder(r)

        type Status struct {
        Status         string `json:"status"`
        Error          string `json:"error"`
        Progress       string `json:"progress"`
        ProgressDetail struct {
            Current int `json:"current"`
            Total   int `json:"total"`
            } `json:"progressDetail"`
        }
	var status *Status

        for {
        if err := fd.Decode(&status); err != nil {
	    if err == io.EOF {
		    break
            }
            fmt.Println("caught ", err, status)
        }
        if status.Error != "" {
		fmt.Println("Failed ", status.Error)
	}
	fmt.Printf("Status: %+v\n", status)
        }
        fmt.Println("done")
```

if the error is caught correctly, done shouldn't be printed.

Reply:
I was thinking about something like:
```go
package main

import (
        "bytes"
        "encoding/json"
        "fmt"
        "io"
        "io/ioutil"
        "os"
        "strings"
)

// ImagePullStatus is the status of the downloaded Image chunk
type ImagePullStatus struct {
        Status         string `json:"status"`
        Error          string `json:"error"`
        Progress       string `json:"progress"`
        ProgressDetail struct {
                Current int `json:"current"`
                Total   int `json:"total"`
        } `json:"progressDetail"`
}

func main() {
        r := ioutil.NopCloser(strings.NewReader("hello world")) // r type is io.ReadCloser

        // Ensure ReadCloser will be closed and error will be handled.
        //
        // Error handling is not critical for execution, so only inform user
        // if error occurs, but proceed with execution.
        defer func() {
                fmt.Println("Closing reader")
                if err := r.Close(); err != nil {
                        fmt.Printf("Closing reader: %v\n", err)
                }
        }()

        // Make a copy of logs produced by reader into buffer, so it can be copied into stdout.
        var bufferRead bytes.Buffer

        reader := io.TeeReader(r, &bufferRead)

        // Stream copied logs.
        //
        // Error handling is not critical for execution, so only inform user
        // if error occurs, but proceed with execution.
        go func() {
                fmt.Println("Copying logs to stdout")

                if _, err := io.Copy(&bufferRead, os.Stdout); err != nil {
                        fmt.Printf("Writing logs to     stdout: %v\n", err)
                }
        }()

        fd := json.NewDecoder(reader)

        var status *ImagePullStatus

        fmt.Println("Entering decode loop")
        for {
                if err := fd.Decode(&status); err != nil {
                        if err == io.EOF {
                                break
                        }

                        fmt.Println("Decoding error occurred: %v", err)
                        continue
                }

                if status.Error != "" {
                        fmt.Printf("Error occurred: %v", status.Error)
                        return
                }
        }
}
```

This loops infinitely, it shouldn't, but I hope you get the idea.

Reply:
Hey! A couple of things! @invidian thank you for catching this. @Cbkhare you should limit the scope of a PR to a minimum. Changing the logging behavior is probably out of scope. 

A few days ago something funny happened. Another member of the community asked for the opposite of what @invidian wants. Let's discuss "image pull logging" there because as I said, this change is out of scope for this PR.

https://github.com/tinkerbell/tink/issues/451